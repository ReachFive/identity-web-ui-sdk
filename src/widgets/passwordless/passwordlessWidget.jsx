import React from 'react';

import { email } from '../../core/validation';

import { createMultiViewWidget } from '../../components/widget/widget';
import { Info, Intro, Separator } from '../../components/miscComponent';

import { createForm } from '../../components/form/formComponent';
import { simpleField } from '../../components/form/fields/simpleField';
import phoneNumberField from '../../components/form/fields/phoneNumberField';
import SocialButtons from '../../components/form/socialButtonsComponent';

const EmailInputForm = createForm({
    prefix: 'r5-passwordless-',
    fields: [
        simpleField({
            key: 'email',
            label: 'email',
            type: 'email',
            validator: email
        })
    ]
});

const phoneNumberInputForm = config => createForm({
    prefix: 'r5-passwordless-sms-',
    fields: [phoneNumberField({ required: true }, config)]
});

const VerificationCodeInputForm = createForm({
    prefix: 'r5-passwordless-sms-',
    fields: [
        simpleField({
            key: 'verification_code',
            label: 'verificationCode',
            type: 'text'
        })
    ]
});

class MainView extends React.Component {
    handleSubmit = data => this.props.apiClient.startPasswordless(data, this.props.auth).then(_ => data);

    handleSuccess = data => {
        return data.email
            ? this.props.goTo('emailSent')
            : this.props.goTo('verificationCode', data);
    };

    render() {
        const { i18n, showSocialLogins, socialProviders, showIntro, authType } = this.props;
        const isEmail = authType === 'magic_link' || authType === '' || !authType;
        const PhoneNumberInputForm = phoneNumberInputForm(this.props.config);

        return <div>
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <SocialButtons providers={socialProviders} auth={this.props.auth} />
            )}
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <Separator text={i18n('or')} />
            )}
            {isEmail && showIntro && <Intro>{i18n('passwordless.intro')}</Intro>}
            {isEmail && <EmailInputForm handler={this.handleSubmit} onSuccess={this.handleSuccess} />}
            {!isEmail && showIntro && <Intro>{i18n('passwordless.sms.intro')}</Intro>}
            {!isEmail && <PhoneNumberInputForm handler={this.handleSubmit} onSuccess={this.handleSuccess} />}
        </div>;
    }
}

class VerificationCodeView extends React.Component {
    handleSubmit = data => {
        const { apiClient, auth, phoneNumber } = this.props;

        return apiClient.verifyPasswordless({ phoneNumber, ...data }, auth);
    };

    render() {
        return <div>
            <Info>{this.props.i18n('passwordless.sms.verification.intro')}</Info>
            <VerificationCodeInputForm handler={this.handleSubmit} />
        </div>;
    }
}
const EmailSentView = ({ i18n }) => <Info>{i18n('passwordless.emailSent')}</Info>;

export default createMultiViewWidget({
    initialView: 'main',
    views: {
        main: MainView,
        emailSent: EmailSentView,
        verificationCode: VerificationCodeView
    },
    prepare: (options, { config }) => ({
        socialProviders: config.socialProviders,
        auth: {},
        showSocialLogins: false,
        showIntro: true,
        authType: 'magic_link',
        ...options
    })
})
