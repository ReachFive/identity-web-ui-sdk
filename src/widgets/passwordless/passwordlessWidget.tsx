import React, { ComponentProps, useLayoutEffect } from 'react';

import { AuthOptions, AuthResult, SingleFactorPasswordlessParams } from '@reachfive/identity-core';

import { CaptchaProvider, WithCaptchaProps, type WithCaptchaToken } from '../../components/captcha';
import phoneNumberField, {
    type PhoneNumberOptions,
} from '../../components/form/fields/phoneNumberField';
import { simpleField } from '../../components/form/fields/simpleField';
import { createForm } from '../../components/form/formComponent';
import { SocialButtons } from '../../components/form/socialButtonsComponent';
import { Info, Intro, Separator } from '../../components/miscComponent';
import { importGoogleRecaptchaScript } from '../../components/reCaptcha';
import { createMultiViewWidget } from '../../components/widget/widget';
import { useI18n } from '../../contexts/i18n';
import { useReachfive } from '../../contexts/reachfive';
import { useRouting } from '../../contexts/routing';
import { email } from '../../core/validation';

import type { Config, OnError, OnSuccess, Prettify } from '../../types';

type EmailFormData = { email: string; captchaToken?: string };

const EmailInputForm = createForm<EmailFormData>({
    prefix: 'r5-passwordless-',
    fields: [
        simpleField({
            key: 'email',
            label: 'email',
            type: 'email',
            validator: email,
        }),
    ],
});

type PhoneNumberFormData = { phoneNumber: string; captchaToken?: string };

const phoneNumberInputForm = (config: Config) =>
    createForm<PhoneNumberFormData, { phoneNumberOptions?: PhoneNumberOptions }>({
        prefix: 'r5-passwordless-sms-',
        fields: ({ phoneNumberOptions }) => [
            phoneNumberField(
                {
                    required: true,
                    ...phoneNumberOptions,
                },
                config
            ),
        ],
    });

type VerificationCodeFormData = { verificationCode: string };

const VerificationCodeInputForm = createForm<VerificationCodeFormData>({
    prefix: 'r5-passwordless-sms-',
    fields: [
        simpleField({
            key: 'verification_code',
            label: 'verificationCode',
            type: 'text',
        }),
    ],
});

interface MainViewProps {
    /**
     * List of authentication options
     */
    auth?: AuthOptions;
    /**
     * The passwordless auth type (`magic_link` or `sms`).
     * @default "magic_link"
     */
    authType?: SingleFactorPasswordlessParams['authType'];
    /**
     * Show the introduction text.
     * @default true
     */
    showIntro?: boolean;
    /**
     * Show the social login buttons.
     * @default false
     */
    showSocialLogins?: boolean;
    /**
     * Lists the available social providers. This is an array of strings.
     *
     * Tip:  If you pass an empty array, social providers will not be displayed.
     */
    socialProviders?: string[];
    /**
     * Phone number field options.
     */
    phoneNumberOptions?: PhoneNumberOptions;
    /**
     * Callback function called when the request has succeed.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
}

const MainView = ({
    auth,
    authType = 'magic_link',
    recaptcha_enabled = false,
    recaptcha_site_key,
    captchaFoxEnabled = false,
    captchaFoxMode = 'hidden',
    captchaFoxSiteKey,
    showIntro = true,
    showSocialLogins = false,
    socialProviders,
    phoneNumberOptions,
    onError = (() => {}) as OnError,
    onSuccess = (() => {}) as OnSuccess,
}: WithCaptchaProps<MainViewProps>) => {
    const { client: coreClient, config } = useReachfive();
    const i18n = useI18n();
    const { goTo } = useRouting();

    useLayoutEffect(() => {
        importGoogleRecaptchaScript(recaptcha_site_key);
    }, [recaptcha_site_key]);

    const callback = (data: WithCaptchaToken<EmailFormData | PhoneNumberFormData>) =>
        coreClient
            .startPasswordless(
                {
                    authType,
                    ...data,
                },
                auth
            )
            .then(() => data);

    const handleSuccess = (data: EmailFormData | PhoneNumberFormData) => {
        onSuccess({ name: 'otp_sent', authType });
        if ('email' in data) {
            goTo('emailSent');
        } else {
            goTo<VerificationCodeViewState>('verificationCode', data);
        }
    };

    const isEmail = authType === 'magic_link';
    const PhoneNumberInputForm = phoneNumberInputForm(config);

    return (
        <div>
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <SocialButtons
                    providers={socialProviders}
                    auth={auth}
                    onSuccess={onSuccess}
                    onError={onError}
                />
            )}
            {showSocialLogins && socialProviders && socialProviders.length > 0 && (
                <Separator text={i18n('or')} />
            )}
            {isEmail && showIntro && <Intro>{i18n('passwordless.intro')}</Intro>}
            {isEmail && (
                <CaptchaProvider
                    recaptcha_enabled={recaptcha_enabled}
                    recaptcha_site_key={recaptcha_site_key}
                    captchaFoxEnabled={captchaFoxEnabled}
                    captchaFoxSiteKey={captchaFoxSiteKey}
                    captchaFoxMode={captchaFoxMode}
                    action="passwordless_email"
                >
                    <EmailInputForm
                        handler={callback}
                        onSuccess={handleSuccess}
                        onError={onError}
                    />
                </CaptchaProvider>
            )}
            {!isEmail && showIntro && <Intro>{i18n('passwordless.sms.intro')}</Intro>}
            {!isEmail && (
                <CaptchaProvider
                    recaptcha_enabled={recaptcha_enabled}
                    recaptcha_site_key={recaptcha_site_key}
                    captchaFoxEnabled={captchaFoxEnabled}
                    captchaFoxSiteKey={captchaFoxSiteKey}
                    captchaFoxMode={captchaFoxMode}
                    action="passwordless_phone"
                >
                    <PhoneNumberInputForm
                        handler={callback}
                        onSuccess={handleSuccess}
                        onError={onError}
                        phoneNumberOptions={phoneNumberOptions}
                    />
                </CaptchaProvider>
            )}
        </div>
    );
};

interface VerificationCodeViewProps {
    /**
     * The passwordless auth type (`magic_link` or `sms`).
     * @default "magic_link"
     */
    authType?: SingleFactorPasswordlessParams['authType'];
    /**
     * Callback function called when the request has succeed.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
}

type VerificationCodeViewState = {
    phoneNumber: string;
};

const VerificationCodeView = ({
    authType = 'magic_link',
    recaptcha_enabled = false,
    recaptcha_site_key,
    captchaFoxEnabled = false,
    captchaFoxMode = 'hidden',
    captchaFoxSiteKey,
    onSuccess = (() => {}) as OnSuccess,
    onError = (() => {}) as OnError,
}: WithCaptchaProps<VerificationCodeViewProps>) => {
    const { client: coreClient } = useReachfive();
    const i18n = useI18n();
    const { params } = useRouting();
    const { phoneNumber } = params as VerificationCodeViewState;

    const handleSubmit = (data: WithCaptchaToken<VerificationCodeFormData>) => {
        return coreClient
            .verifyPasswordless({
                authType,
                phoneNumber,
                ...data,
            })
            .then(result => {
                if (AuthResult.isAuthResult(result)) {
                    onSuccess({
                        name: 'login',
                        authResult: result,
                        authType,
                        identifierType: 'phone_number',
                    });
                } else {
                    onError();
                }
            })
            .catch(onError);
    };

    return (
        <div>
            <CaptchaProvider
                recaptcha_enabled={recaptcha_enabled}
                recaptcha_site_key={recaptcha_site_key}
                captchaFoxEnabled={captchaFoxEnabled}
                captchaFoxSiteKey={captchaFoxSiteKey}
                captchaFoxMode={captchaFoxMode}
                action="verify_passwordless_sms"
            >
                <Info>{i18n('passwordless.sms.verification.intro')}</Info>
                <VerificationCodeInputForm handler={handleSubmit} onError={onError} />
            </CaptchaProvider>
        </div>
    );
};

const EmailSentView = () => {
    const i18n = useI18n();
    return <Info>{i18n('passwordless.emailSent')}</Info>;
};

export type PasswordlessWidgetProps = Prettify<
    ComponentProps<typeof MainView> & ComponentProps<typeof VerificationCodeView>
>;

export default createMultiViewWidget<PasswordlessWidgetProps>({
    initialView: 'main',
    views: {
        main: MainView,
        emailSent: EmailSentView,
        verificationCode: VerificationCodeView,
    },
    prepare: (options, { config }) => ({
        socialProviders: config.socialProviders,
        ...options,
    }),
});
