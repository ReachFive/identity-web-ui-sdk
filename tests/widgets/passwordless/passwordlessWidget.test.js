import renderer from 'react-test-renderer';
import 'jest-styled-components';
import $ from 'cheerio';
import { render } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import passwordlessWidget from '../../../src/widgets/passwordless/passwordlessWidget'

Enzyme.configure({ adapter: new Adapter() });

const textFilter = expected => (i, el) => $(el).text() === expected;

const defaultConfig = { domain: 'local.reach5.net' };

describe('Snapshot', () => {
    const generateSnapshot = (options, config = defaultConfig) => () => {
        const tree = passwordlessWidget(options, { config, apiClient: {} })
            .then(result => renderer.create(result).toJSON());

        expect(tree).resolves.toMatchSnapshot();
    };

    describe('passwordless', () => {
        test('default', generateSnapshot());

        test('no intro', generateSnapshot({ showIntro: false }));

        test('sms', generateSnapshot({ authType: "sms" }));
    });
});

describe('DOM testing', () => {
    const generateComponent = async (options, config = defaultConfig) => {
        const result = await passwordlessWidget(options, { config, apiClient: {} });

        return render(result);
    };

    describe('passwordless', () => {
        test('default', async () => {
            expect.assertions(4);
            const instance = await generateComponent({});

            // Intro
            expect(
                instance.find('div').filter(textFilter('passwordless.intro'))
            ).toHaveLength(1);

            // Label
            expect(instance.find('label').text()).toBe('email');

            // Input email
            expect(instance.find('[type="email"]')).toHaveLength(1);

            // Form button
            expect(instance.find('button').text()).toBe('send');
        });

        test('no intro', async () => {
            expect.assertions(1);
            const instance = await generateComponent({ showIntro: false });

            // Intro
            expect(
                instance.find('div').filter(textFilter('passwordless.intro'))
            ).toHaveLength(0);
        });

        test('by phone number', async () => {
            expect.assertions(4);
            const instance = await generateComponent({ authType: "sms" });

            // Intro
            expect(
                instance.find('div').filter(textFilter('passwordless.sms.intro'))
            ).toHaveLength(1);

            // Label
            expect(instance.find('label').text()).toBe('phoneNumber');

            // Input phone number
            expect(instance.find('[type="tel"]')).toHaveLength(1);

            // Form button
            expect(instance.find('button').text()).toBe('send');
        });
    });
});
