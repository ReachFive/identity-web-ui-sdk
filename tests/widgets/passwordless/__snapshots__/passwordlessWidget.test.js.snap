// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot passwordless default 1`] = `
.c1 {
  text-align: center;
  margin-bottom: 10px;
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c6 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c6:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c6:hover,
.c6:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c6[disabled] {
  opacity: .65;
}

.c2 {
  position: relative;
}

.c4 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c3 {
  margin-bottom: 10px;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
}

.c5:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c5:disabled,
.c5[readonly] {
  background-color: #e9ecef;
}

.c5::-webkit-input-placeholder {
  color: #868e96;
}

.c5::-moz-placeholder {
  color: #868e96;
}

.c5:-ms-input-placeholder {
  color: #868e96;
}

.c5::placeholder {
  color: #868e96;
}

<div
  className="c0"
>
  <div>
    <div
      className="c1"
    >
      Enter your email to sign in or create an account.
    </div>
    <form
      className="c2"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="c3"
      >
        <label
          className="c4"
          htmlFor="email"
        >
          Email address
        </label>
        <input
          className="c5"
          data-testid="email"
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Email address"
          required={true}
          title="Email address"
          type="email"
          value=""
        />
      </div>
      <button
        className="c6"
        data-testid="submit"
        disabled={false}
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</div>
`;

exports[`Snapshot passwordless no intro 1`] = `
.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c5:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c5:hover,
.c5:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c5[disabled] {
  opacity: .65;
}

.c1 {
  position: relative;
}

.c3 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c2 {
  margin-bottom: 10px;
}

.c4 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
}

.c4:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c4:disabled,
.c4[readonly] {
  background-color: #e9ecef;
}

.c4::-webkit-input-placeholder {
  color: #868e96;
}

.c4::-moz-placeholder {
  color: #868e96;
}

.c4:-ms-input-placeholder {
  color: #868e96;
}

.c4::placeholder {
  color: #868e96;
}

<div
  className="c0"
>
  <div>
    <form
      className="c1"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="c2"
      >
        <label
          className="c3"
          htmlFor="email"
        >
          Email address
        </label>
        <input
          className="c4"
          data-testid="email"
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Email address"
          required={true}
          title="Email address"
          type="email"
          value=""
        />
      </div>
      <button
        className="c5"
        data-testid="submit"
        disabled={false}
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</div>
`;

exports[`Snapshot passwordless sms 1`] = `
.c1 {
  text-align: center;
  margin-bottom: 10px;
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c6 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c6:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c6:hover,
.c6:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c6[disabled] {
  opacity: .65;
}

.c2 {
  position: relative;
}

.c4 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c3 {
  margin-bottom: 10px;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
}

.c5:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c5:disabled,
.c5[readonly] {
  background-color: #e9ecef;
}

.c5::-webkit-input-placeholder {
  color: #868e96;
}

.c5::-moz-placeholder {
  color: #868e96;
}

.c5:-ms-input-placeholder {
  color: #868e96;
}

.c5::placeholder {
  color: #868e96;
}

<div
  className="c0"
>
  <div>
    <div
      className="c1"
    >
      Enter your phone number to sign in or create an account.
    </div>
    <form
      className="c2"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="c3"
      >
        <label
          className="c4"
          htmlFor="phone_number"
        >
          Phone number
        </label>
        <input
          className="c5"
          data-testid="phone_number"
          id="phone_number"
          name="phone_number"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Phone number"
          required={true}
          title="Phone number"
          type="tel"
          value=""
        />
      </div>
      <button
        className="c6"
        data-testid="submit"
        disabled={false}
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</div>
`;
